lib/cretonne/src/loop_analysis.rs

```diff
@@ -0,0 +1,324 @@
+//! A loop analysis respresented as mappings of loops to their header Ebb
+//! and parent in the loop tree.
+
+
+use ir::{Function, Ebb, Layout, Loop};
+use flowgraph::ControlFlowGraph;
+use dominator_tree::DominatorTree;
+use entity_map::{EntityMap, PrimaryEntityData};
+use packed_option::PackedOption;
+use std::collections::HashSet;
+
+/// Loop tree information for a single function.
+///
+/// Loops are referenced by the Loop object, and for each loop you can access its header EBB,
+/// its eventual parent in the loop tree and all the EBB belonging to the loop.
+pub struct LoopAnalysis {
+    loops: EntityMap<Loop, LoopData>,
+    ebbs: EntityMap<Ebb, PackedOption<Loop>>,
+}
+
+struct LoopData {
+    header: Ebb,
+    parent: PackedOption<Loop>,
+}
+
+impl PrimaryEntityData for LoopData {}
+
+/// Methods for querying the loop analysis.
+impl LoopAnalysis {
+    /// Allocate a new blank loop analysis struct. Use `compute` to compute the loop analysis for
+    /// a function.
+    pub fn new() -> LoopAnalysis {
+        LoopAnalysis {
+            loops: EntityMap::new(),
+            ebbs: EntityMap::new(),
+        }
+    }
+
+    /// Returns all the loops contained in a function.
+    pub fn get_loops(&self) -> Vec<Loop> {
+        self.loops.keys().collect()
+    }
+
+    /// Returns the header EBB of a particular loop.
+    ///
+    /// The characteristic property of a loop header block is that it dominates some of its
+    /// predecessors.
+    pub fn get_loop_header(&self, lp: Loop) -> Ebb {
+        self.loops[lp].header
+    }
+
+    /// Return the eventual parents of a loop in the loop tree.
+    pub fn get_loop_parent(&self, lp: Loop) -> Option<Loop> {
+        self.loops[lp].parent.expand()
+    }
+
+    /// Determine if an Ebb belongs to a loop by running a finger along the loop tree.
+    ///
+    /// Returns `true` if `ebb` is in loop `lp`.
+    pub fn is_in_loop(&self, ebb: Ebb, lp: Loop) -> bool {
+        let mut ebb_loop = self.ebbs[ebb].expand();
+        while ebb_loop != None {
+            if ebb_loop.unwrap() == lp {
+                return true;
+            } else {
+                ebb_loop = self.get_loop_parent(ebb_loop.unwrap());
+            }
+        }
+        false
+    }
+
+    /// Return ebbs from a loop in post-order, starting from an entry point in the block.
+    pub fn postorder_ebbs_loop(&self, cfg: &ControlFlowGraph, lp: Loop) -> Vec<Ebb> {
```

Comment:
I'm a little skeptical about this method, it's throwing around a lot of big data structures.

Perhaps you could move it into the LICM module? I don't think it would be used by anything else.